src/audio/component.c

```diff
@@ -154,6 +154,19 @@ void comp_get_copy_limits(struct comp_buffer *source, struct comp_buffer *sink,
 	cl->sink_bytes = cl->frames * cl->sink_frame_bytes;
 }
 
+void comp_get_copy_limits_aligned(struct comp_buffer *source, struct comp_buffer *sink,
+				  struct comp_copy_limits *cl, uint32_t num_power_of_2)
```

Comment:
can we use bytes for alignment as it's easier to follow as other similar APIs use bytes.

Reply:
@singalsu Is the processing unit frame？

Reply:
no, bytes is used by similar APIs (for allocation) and is also whats' used by the SIMD instruction (as 8 in this case). Some time it will nicely align with frame size e.g. 32 bit stereo stream has frame size of 8 bytes.

So in the volume use case where we need 8 byte alignment we would call
```c
comp_get_copy_limits_aligned(source, sink, &cl, 8);
```
This would give us the copy limits where the **start** address and **size** are aligned to 8 bytes.

Reply:
We need for SIMD frames align, maybe have both as input parameter?

Reply:
@lgirdwood Any example where bytes count align would be beneficial?

Reply:
but almost all the components are processed by frame, even if we align bytes, the byte length to be processed should correspond to the length of an integer frame. I am concerned about the cost.

Reply:
or can we use comp_get_copy_limits_frame_aligned(source, sink, &cl, frame_num) to avoid confusion？

Reply:
We can remove the alignment here and optimize further by calculating it only once at stream config (instead of every copy()) e.g.
```diff
diff --git a/src/include/sof/audio/audio_stream.h b/src/include/sof/audio/audio_stream.h
index 267022df5..296e95578 100644
--- a/src/include/sof/audio/audio_stream.h
+++ b/src/include/sof/audio/audio_stream.h
@@ -59,6 +59,7 @@ struct audio_stream {
 
        uint32_t rate;          /**< Number of data frames per second [Hz] */
        uint16_t channels;      /**< Number of samples in each frame */
+       uint16_t frame_align;   /**< Frame alignment (power of 2) */
 
        bool overrun_permitted; /**< indicates whether overrun is permitted */
        bool underrun_permitted; /**< indicates whether underrun is permitted */

```
We keep a local per stream copy of the alignment and set it once at stream config. We can then optimize and remove some divisions.
i.e. your patch calls audio_stream_avail_frames() 
```c
static inline uint32_t
audio_stream_get_avail_frames(const struct audio_stream *stream)
{
	return audio_stream_get_avail_bytes(stream) /
		audio_stream_frame_bytes(stream);
}
```
which could be replaced with
```c
static inline uint32_t
audio_stream_get_avail_frames_aligned(const struct audio_stream *stream)
{
	return audio_stream_get_avail_bytes(stream) & stream->frame_align;
}
```
Which removes the division too.

Reply:
`static inline uint32_t
audio_stream_get_avail_frames_aligned(const struct audio_stream *stream)
{
	return audio_stream_get_avail_bytes(stream) & stream->frame_align;
}`
do you mean return audio_stream_get_avail_frames(stream) &（~( stream->frame_align - 1)）?

Reply:
@andrula-song yes, I mean we can set the alignment value (since it will be constant) during the stream init and pre calculate the mask ```（~( stream->frame_align - 1)）```.
If we also enforce that alignment will always be a power of 2 then the conversion from bytes(aligned) to frames(aligned) should be a shift and a subtraction instead of a division (please check though).